<a href="https://colab.research.google.com/github/DiwakarBasnet/Machine-Learning/blob/main/Naive_Bayes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Naive Bayes classification

If we have hundreds of thousands of data points and quite a few variables in our training data set and we want to see first cut of model within an hour. In such a situation, we could use Naive Bayes which can be extremley fast relative to other classification algorithms.

It is a classification technique based on Bayes theorem with an assumption of independence between attributes of data points. From Naive Bayes theorem:-

\begin{align}
  P(C \mid X) = \frac{P(X \mid C) \, P(C)}{P(X)}
\end{align}

\begin{align}
  P(C \mid X) = P(x_1 \mid C)*P(x_2 \mid C)*P(x_3 \mid C)*.....*P(x_n \mid C) * P(C)
\end{align}

When assumption of independence holds, a Naive Bayes classifier performs better compared to other model like logistic regression.

# Data Preprocessing

In [1]:
# Importing libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

%matplotlib inline

In [2]:
# Import dataset
url = 'https://raw.githubusercontent.com/DiwakarBasnet/Machine-Learning/main/chapter-02-supervised-learning/dataset/adult.csv?token=GHSAT0AAAAAABYAO27LRLHQGD5RELKSKXDYY2OPXAQ'
df = pd.read_csv(url, header=None, sep=',\s')

/usr/local/lib/python3.7/dist-packages/pandas/util/_decorators.py:311: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  return func(*args, **kwargs)


##Exploratory data analysis

In [3]:
df.shape

(32561, 15)

In [4]:
df.head()

0                 1       2          3   4                   5   \
0  39         State-gov   77516  Bachelors  13       Never-married   
1  50  Self-emp-not-inc   83311  Bachelors  13  Married-civ-spouse   
2  38           Private  215646    HS-grad   9            Divorced   
3  53           Private  234721       11th   7  Married-civ-spouse   
4  28           Private  338409  Bachelors  13  Married-civ-spouse   

                  6              7      8       9     10  11  12  \
0       Adm-clerical  Not-in-family  White    Male  2174   0  40   
1    Exec-managerial        Husband  White    Male     0   0  13   
2  Handlers-cleaners  Not-in-family  White    Male     0   0  40   
3  Handlers-cleaners        Husband  Black    Male     0   0  40   
4     Prof-specialty           Wife  Black  Female     0   0  40   

              13     14  
0  United-States  <=50K  
1  United-States  <=50K  
2  United-States  <=50K  
3  United-States  <=50K  
4           Cuba  <=50K

In [5]:
# Renaming the columns
col_names = ['age', 'workclass', 'fnlwgt', 'education', 'education_num', 'marital_status', 'occupation', 'relationship',
             'race', 'sex', 'capital_gain', 'capital_loss', 'hours_per_week', 'native_country', 'income']

df.columns = col_names
df.head()

age         workclass  fnlwgt  education  education_num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital_status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital_gain  capital_loss  hours_per_week native_country income  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K

In [6]:
# Summary of dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education_num   32561 non-null  int64 
 5   marital_status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital_gain    32561 non-null  int64 
 11  capital_loss    32561 non-null  int64 
 12  hours_per_week  32561 non-null  int64 
 13  native_country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


###Explore Categorical Varaibles

In [7]:
# find categorical variables
categorical = [var for var in df.columns if df[var].dtype == 'O']

print("There are {} categorical variables\n".format(len(categorical)))
print("The categorical variables are:\n",categorical)

There are 9 categorical variables

The categorical variables are:
 ['workclass', 'education', 'marital_status', 'occupation', 'relationship', 'race', 'sex', 'native_country', 'income']


In [8]:
df[categorical].head()

workclass  education      marital_status         occupation  \
0         State-gov  Bachelors       Never-married       Adm-clerical   
1  Self-emp-not-inc  Bachelors  Married-civ-spouse    Exec-managerial   
2           Private    HS-grad            Divorced  Handlers-cleaners   
3           Private       11th  Married-civ-spouse  Handlers-cleaners   
4           Private  Bachelors  Married-civ-spouse     Prof-specialty   

    relationship   race     sex native_country income  
0  Not-in-family  White    Male  United-States  <=50K  
1        Husband  White    Male  United-States  <=50K  
2  Not-in-family  White    Male  United-States  <=50K  
3        Husband  Black    Male  United-States  <=50K  
4           Wife  Black  Female           Cuba  <=50K

In [9]:
# Frequency counts of values in categorical variables
for var in categorical:
  print(df[var].value_counts())

Private             22696
Self-emp-not-inc     2541
Local-gov            2093
?                    1836
State-gov            1298
Self-emp-inc         1116
Federal-gov           960
Without-pay            14
Never-worked            7
Name: workclass, dtype: int64
HS-grad         10501
Some-college     7291
Bachelors        5355
Masters          1723
Assoc-voc        1382
11th             1175
Assoc-acdm       1067
10th              933
7th-8th           646
Prof-school       576
9th               514
12th              433
Doctorate         413
5th-6th           333
1st-4th           168
Preschool          51
Name: education, dtype: int64
Married-civ-spouse       14976
Never-married            10683
Divorced                  4443
Separated                 1025
Widowed                    993
Married-spouse-absent      418
Married-AF-spouse           23
Name: marital_status, dtype: int64
Prof-specialty       4140
Craft-repair         4099
Exec-managerial      4066
Adm-clerical         377

Generally, the missing values are coded as **NAN** and python will detect them with the usual command of df.isnull().sum() or simply df.info(). But, in this case the missing value are coded as ?. Python fail to detect these as missing values because it do not consider ? as missing value. SO, now we have to replace ? with NAN so that python can detect these missing values.

In [10]:
# replacing ? with NaN
for x in categorical:
  df[x].replace('?', np.NaN, inplace=True)

In [11]:
df.workclass.value_counts()

Private             22696
Self-emp-not-inc     2541
Local-gov            2093
State-gov            1298
Self-emp-inc         1116
Federal-gov           960
Without-pay            14
Never-worked            7
Name: workclass, dtype: int64

In [12]:
# Check missing value in categorical variables
df[categorical].isnull().sum()

workclass         1836
education            0
marital_status       0
occupation        1843
relationship         0
race                 0
sex                  0
native_country     583
income               0
dtype: int64

The number of labels within a categorical variable is known as **cardinality**. A high number of labels within a variable is known as **high cardinality**. High cardinality may pose some serious problems in the machine learning model. So, we will check for high cardinality.

In [13]:
# Check cardinality in categorical variables
for var in categorical:
  print(var,' contains ',len(df[var].unique()),' labels')

workclass  contains  9  labels
education  contains  16  labels
marital_status  contains  7  labels
occupation  contains  15  labels
relationship  contains  6  labels
race  contains  5  labels
sex  contains  2  labels
native_country  contains  42  labels
income  contains  2  labels


native_country column contains relatively large number of labels as compared to other columns.

###Explore Numerical Variables

In [14]:
numerical = [var for var in df.columns if df[var].dtype != 'O']
print('There are {} numerical variables.\n'.format(len(numerical)))
print('The numerical variables are:\n',numerical)

There are 6 numerical variables.

The numerical variables are:
 ['age', 'fnlwgt', 'education_num', 'capital_gain', 'capital_loss', 'hours_per_week']


In [15]:
df[numerical].head()

age  fnlwgt  education_num  capital_gain  capital_loss  hours_per_week
0   39   77516             13          2174             0              40
1   50   83311             13             0             0              13
2   38  215646              9             0             0              40
3   53  234721              7             0             0              40
4   28  338409             13             0             0              40

In [16]:
# Check missing values in numerical variables
df[numerical].isnull().sum()

age               0
fnlwgt            0
education_num     0
capital_gain      0
capital_loss      0
hours_per_week    0
dtype: int64

##Train test split

In [17]:
X = df.drop(['income'], axis=1)   # features
y = df['income']                  # target

In [18]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)

In [19]:
X_train.shape, X_test.shape

((22792, 14), (9769, 14))

##Feature Engineering

Feateure Engineering is the process of transforming raw data into useful feaetures that help us to understand our model better and increase its predictive power.

In [20]:
# display categorical variables
categorical = [col for col in X_train.columns if X_train[col].dtypes == 'O']
categorical

['workclass',
 'education',
 'marital_status',
 'occupation',
 'relationship',
 'race',
 'sex',
 'native_country']

In [21]:
# display numerical variables
numerical = [col for col in X_train.columns if X_train[col].dtypes != 'O']
numerical

['age',
 'fnlwgt',
 'education_num',
 'capital_gain',
 'capital_loss',
 'hours_per_week']

In [22]:
# print categorical variables with missing data
for col in categorical:
  if X_train[col].isnull().mean() > 0:
    print(col, (X_train[col].isnull().mean()))

workclass 0.055984555984555984
occupation 0.05607230607230607
native_country 0.018164268164268166


In [23]:
# input missing categorical variables with most frequent values
for df2 in [X_train, X_test]:
  df2['workclass'].fillna(X_train['workclass'].mode()[0], inplace=True)
  df2['occupation'].fillna(X_train['occupation'].mode()[0], inplace=True)
  df2['native_country'].fillna(X_train['native_country'].mode()[0], inplace=True)

In [24]:
# check missing value in categorical variable
X_train[categorical].isnull().sum()

workclass         0
education         0
marital_status    0
occupation        0
relationship      0
race              0
sex               0
native_country    0
dtype: int64

In [25]:
X_test[categorical].isnull().sum()

workclass         0
education         0
marital_status    0
occupation        0
relationship      0
race              0
sex               0
native_country    0
dtype: int64

###Encode categorical variables

In [26]:
!pip install category_encoders

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 72 kB 628 kB/s 


In [27]:
X_train[categorical].head()

workclass     education      marital_status    occupation  \
32098    Private       HS-grad  Married-civ-spouse  Craft-repair   
25206  State-gov       HS-grad            Divorced  Adm-clerical   
23491    Private  Some-college  Married-civ-spouse         Sales   
12367    Private       HS-grad       Never-married  Craft-repair   
7054     Private       7th-8th       Never-married  Craft-repair   

        relationship   race     sex native_country  
32098        Husband  White    Male  United-States  
25206      Unmarried  White  Female  United-States  
23491        Husband  White    Male  United-States  
12367  Not-in-family  White    Male      Guatemala  
7054   Not-in-family  White    Male        Germany

In [28]:
# Encode categorical variables with one hot encoding
import category_encoders as ce

encoder = ce.OneHotEncoder(cols = categorical)
new_X_train = encoder.fit_transform(X_train)
new_X_test = encoder.transform(X_test)

In [30]:
new_X_test.head()

age  workclass_1  workclass_2  workclass_3  workclass_4  workclass_5  \
22278   27            1            0            0            0            0   
8950    27            1            0            0            0            0   
7838    25            1            0            0            0            0   
16505   46            1            0            0            0            0   
19140   45            1            0            0            0            0   

       workclass_6  workclass_7  workclass_8  fnlwgt  ...  native_country_32  \
22278            0            0            0  177119  ...                  0   
8950             0            0            0  216481  ...                  0   
7838             0            0            0  256263  ...                  0   
16505            0            0            0  147640  ...                  0   
19140            0            0            0  172822  ...                  0   

       native_country_33  native_country_34  native_country_35  \
22278                  0                  0                  0   
8950                   0                  0                  0   
7838                   0                  0                  0   
16505                  0                  0                  0   
19140                  0                  0                  0   

       native_country_36  native_country_37  native_country_38  \
22278                  0                  0                  0   
8950                   0                  0                  0   
7838                   0                  0                  0   
16505                  0                  0                  0   
19140                  0                  0                  0   

       native_country_39  native_country_40  native_country_41  
22278                  0                  0                  0  
8950                   0                  0                  0  
7838                   0                  0                  0  
16505                  0                  0                  0  
19140                  0                  0                  0  

[5 rows x 105 columns]

In [29]:
new_X_train.head()

age  workclass_1  workclass_2  workclass_3  workclass_4  workclass_5  \
32098   45            1            0            0            0            0   
25206   47            0            1            0            0            0   
23491   48            1            0            0            0            0   
12367   29            1            0            0            0            0   
7054    23            1            0            0            0            0   

       workclass_6  workclass_7  workclass_8  fnlwgt  ...  native_country_32  \
32098            0            0            0  170871  ...                  0   
25206            0            0            0  108890  ...                  0   
23491            0            0            0  187505  ...                  0   
12367            0            0            0  145592  ...                  0   
7054             0            0            0  203003  ...                  0   

       native_country_33  native_country_34  native_country_35  \
32098                  0                  0                  0   
25206                  0                  0                  0   
23491                  0                  0                  0   
12367                  0                  0                  0   
7054                   0                  0                  0   

       native_country_36  native_country_37  native_country_38  \
32098                  0                  0                  0   
25206                  0                  0                  0   
23491                  0                  0                  0   
12367                  0                  0                  0   
7054                   0                  0                  0   

       native_country_39  native_country_40  native_country_41  
32098                  0                  0                  0  
25206                  0                  0                  0  
23491                  0                  0                  0  
12367                  0                  0                  0  
7054                   0                  0                  0  

[5 rows x 105 columns]

In [31]:
print(new_X_train.shape,', ',new_X_test.shape)

(22792, 105) ,  (9769, 105)


##Feature Scaling

In [32]:
cols = new_X_train.columns

In [33]:
from sklearn.preprocessing import RobustScaler

scaler = RobustScaler()

X_train = scaler.fit_transform(new_X_train)
X_test = scaler.transform(new_X_test)

In [34]:
X_train

array([[ 0.4 ,  0.  ,  0.  , ...,  0.  ,  0.  ,  0.  ],
       [ 0.5 , -1.  ,  1.  , ...,  0.  ,  0.  ,  0.  ],
       [ 0.55,  0.  ,  0.  , ...,  0.  ,  0.  ,  0.  ],
       ...,
       [-0.7 ,  0.  ,  0.  , ...,  0.  ,  0.  ,  0.  ],
       [ 0.4 , -1.  ,  0.  , ...,  0.  ,  0.  ,  0.  ],
       [-0.6 ,  0.  ,  0.  , ...,  0.  ,  0.  ,  0.  ]])

In [35]:
X_train = pd.DataFrame(X_train, columns=[cols])
X_test = pd.DataFrame(X_test, columns=[cols])

X_train.head()

age workclass_1 workclass_2 workclass_3 workclass_4 workclass_5  \
0  0.40         0.0         0.0         0.0         0.0         0.0   
1  0.50        -1.0         1.0         0.0         0.0         0.0   
2  0.55         0.0         0.0         0.0         0.0         0.0   
3 -0.40         0.0         0.0         0.0         0.0         0.0   
4 -0.70         0.0         0.0         0.0         0.0         0.0   

  workclass_6 workclass_7 workclass_8    fnlwgt  ... native_country_32  \
0         0.0         0.0         0.0 -0.058906  ...               0.0   
1         0.0         0.0         0.0 -0.578076  ...               0.0   
2         0.0         0.0         0.0  0.080425  ...               0.0   
3         0.0         0.0         0.0 -0.270650  ...               0.0   
4         0.0         0.0         0.0  0.210240  ...               0.0   

  native_country_33 native_country_34 native_country_35 native_country_36  \
0               0.0               0.0               0.0               0.0   
1               0.0               0.0               0.0               0.0   
2               0.0               0.0               0.0               0.0   
3               0.0               0.0               0.0               0.0   
4               0.0               0.0               0.0               0.0   

  native_country_37 native_country_38 native_country_39 native_country_40  \
0               0.0               0.0               0.0               0.0   
1               0.0               0.0               0.0               0.0   
2               0.0               0.0               0.0               0.0   
3               0.0               0.0               0.0               0.0   
4               0.0               0.0               0.0               0.0   

  native_country_41  
0               0.0  
1               0.0  
2               0.0  
3               0.0  
4               0.0  

[5 rows x 105 columns]

In [36]:
X_test.head()

age workclass_1 workclass_2 workclass_3 workclass_4 workclass_5  \
0 -0.50         0.0         0.0         0.0         0.0         0.0   
1 -0.50         0.0         0.0         0.0         0.0         0.0   
2 -0.60         0.0         0.0         0.0         0.0         0.0   
3  0.45         0.0         0.0         0.0         0.0         0.0   
4  0.40         0.0         0.0         0.0         0.0         0.0   

  workclass_6 workclass_7 workclass_8    fnlwgt  ... native_country_32  \
0         0.0         0.0         0.0 -0.006571  ...               0.0   
1         0.0         0.0         0.0  0.323136  ...               0.0   
2         0.0         0.0         0.0  0.656361  ...               0.0   
3         0.0         0.0         0.0 -0.253496  ...               0.0   
4         0.0         0.0         0.0 -0.042564  ...               0.0   

  native_country_33 native_country_34 native_country_35 native_country_36  \
0               0.0               0.0               0.0               0.0   
1               0.0               0.0               0.0               0.0   
2               0.0               0.0               0.0               0.0   
3               0.0               0.0               0.0               0.0   
4               0.0               0.0               0.0               0.0   

  native_country_37 native_country_38 native_country_39 native_country_40  \
0               0.0               0.0               0.0               0.0   
1               0.0               0.0               0.0               0.0   
2               0.0               0.0               0.0               0.0   
3               0.0               0.0               0.0               0.0   
4               0.0               0.0               0.0               0.0   

  native_country_41  
0               0.0  
1               0.0  
2               0.0  
3               0.0  
4               0.0  

[5 rows x 105 columns]

#Model training

In [37]:
# train a Gaussian Naive Bayes classifier on the training set
from sklearn.naive_bayes import GaussianNB

gnb = GaussianNB()
gnb.fit(X_train, y_train)

/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:1692: FutureWarning: Feature names only support names that are all strings. Got feature names with dtypes: ['tuple']. An error will be raised in 1.2.
  FutureWarning,


GaussianNB()

#Predict the results

In [38]:
y_pred = gnb.predict(X_test)
y_pred

/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:1692: FutureWarning: Feature names only support names that are all strings. Got feature names with dtypes: ['tuple']. An error will be raised in 1.2.
  FutureWarning,


array(['<=50K', '<=50K', '>50K', ..., '>50K', '<=50K', '<=50K'],
      dtype='<U5')

##Accuracy Score

In [40]:
from sklearn.metrics import accuracy_score

print('Model accuracy score: {0:0.4f}'.format(accuracy_score(y_test, y_pred)))

Model accuracy score: 0.8083


##Overfitting and underfitting

In [41]:
# Scores of training and test set
print('Trainig set score: {:.4f}'.format(gnb.score(X_train, y_train)))
print('Test set score: {:.4f}'.format(gnb.score(X_test, y_test)))

Trainig set score: 0.8067
Test set score: 0.8083


/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:1692: FutureWarning: Feature names only support names that are all strings. Got feature names with dtypes: ['tuple']. An error will be raised in 1.2.
  FutureWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:1692: FutureWarning: Feature names only support names that are all strings. Got feature names with dtypes: ['tuple']. An error will be raised in 1.2.
  FutureWarning,


The training-set accuracy score is 0.8067 while test-set accuracy score is 0.8083. These two values are quite comparable. So, there is no sign of overfitting.

**Compare model accuracy with null accuracy:** The model accuracy is 0.8083, but we cannot say that our model is very good based on the above accuracy. We must compare it with **null accuracy**. Null accuracy is the accuracy that could be achieved by always predicting the most frequent class. So, we should first check the class distribution in the test set.

In [42]:
# Check class distribution in test set
y_test.value_counts()

<=50K    7407
>50K     2362
Name: income, dtype: int64

We can see that the occurance of most frequent class is 7407. So, we can calculate null accuracy by dividing 7407 by total number of occurences.

In [43]:
# check null accuracy score
null_accuracy = 7407/(7407+2362)
print('Null accuracy: {0:0.4f}'.format(null_accuracy))

Null accuracy: 0.7582


We can see that our model accuracy score is 0.8083 but null accuracy score is 0.7582. So, we can conclude that our Gaussian Naive Bayes Classification model is doing a very good job in predicting the class labels.

But, it does not give the underlying distribution of values. Also, it does not tell anything about the type of errors our classifier is making.